pkg/apiserver/profiling/profile.go

```diff
@@ -15,64 +15,169 @@ package profiling
+}
+
+type fetcher struct {
+	ctx        context.Context
+	httpClient *http.Client
+	target     *utils.RequestTargetNode
+	output     string
+	tls        bool
+}
+
+func (f *fetcher) Fetch(src string, duration, timeout time.Duration) (*profile.Profile, string, error) {
+	var url string
+	secs := strconv.Itoa(int(duration / time.Second))
+	switch f.target.Kind {
+	case utils.NodeKindPD:
+		url = "/pd/api/v1/debug/pprof/profile?seconds=" + secs
+	case utils.NodeKindTiDB:
+		url = "/debug/pprof/profile?seconds=" + secs
+	default:
+		return nil, "", fmt.Errorf("unsupported target %s", f.target)
+	}
+	schema := "http"
+	if f.tls {
+		schema = "https"
+	}
+	url = fmt.Sprintf("%s://%s:%d%s", schema, f.target.IP, f.target.Port, url)
+
+	p, err := f.getProfile(f.target, url)
+	return p, url, err
+}
+
+func (f *fetcher) getProfile(target *utils.RequestTargetNode, source string) (*profile.Profile, error) {
+	req, err := http.NewRequest(http.MethodGet, source, nil)
+	if err != nil {
+		return nil, err
+	}
+	req = req.WithContext(f.ctx)
 	if target.Kind == utils.NodeKindPD {
 		// forbidden PD follower proxy
 		req.Header.Add("PD-Allow-follower-handle", "true")
 	}
-	resp, err := httpClient.Do(req)
+	resp, err := f.httpClient.Do(req)
 	if err != nil {
-		return "", fmt.Errorf("failed to send profiling request to %s (url = %s): %s", target, url, err)
+		return nil, err
 	}
-	defer resp.Body.Close()
 	if resp.StatusCode != http.StatusOK {
-		return "", fmt.Errorf("failed to profile %s: status code %s", target, resp.Status)
-	}
-
-	svgFilePath, err := writeProfilingSVG(target, resp.Body, fileNameWithoutExt)
-	if err != nil {
-		return "", fmt.Errorf("failed to write profiling result: %s", err)
+		defer resp.Body.Close()
+		return nil, fmt.Errorf("failed to profile %s: status code %s", target, resp.Status)
 	}
-	return svgFilePath, nil
+	return profile.Parse(resp.Body)
 }
 
-func getProfilingURL(target *utils.RequestTargetNode, profileDurationSecs uint, tls bool) (string, error) {
-	var url string
-	secs := strconv.Itoa(int(profileDurationSecs))
+func profileAndWriteSVG(ctx context.Context, target *utils.RequestTargetNode, fileNameWithoutExt string, profileDurationSecs uint, httpClient *http.Client, tls bool) (string, error) {
 	switch target.Kind {
-	case utils.NodeKindPD:
-		url = "/pd/api/v1/debug/pprof/profile?seconds=" + secs
-	case utils.NodeKindTiKV, utils.NodeKindTiDB:
-		url = "/debug/pprof/profile?seconds=" + secs
+	case utils.NodeKindTiKV:
+		return fetchFlameGraphSVG(ctx, httpClient, target, fileNameWithoutExt, profileDurationSecs, tls)
+	case utils.NodeKindPD, utils.NodeKindTiDB:
+		return fetchPprofSVG(ctx, httpClient, target, fileNameWithoutExt, "dot", profileDurationSecs, tls)
 	default:
 		return "", fmt.Errorf("unsupported target %s", target)
 	}
+}
+
+func fetchFlameGraphSVG(ctx context.Context, httpClient *http.Client, target *utils.RequestTargetNode, fileNameWithoutExt string, profileDurationSecs uint, tls bool) (string, error) {
```

Comment:
maybe we should call it something like fetchTiKVProfiling since this function will not work for all components